Remove variable shadowing in TensorPipe lambda

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39126 Remove variable shadowing in TensorPipe lambda**

`futureResponseMessage` is shadowed in the `pipeWrite` lambda which
creates some confusion, since it is used in the initial error handling but then a future of the same name is created when marking the future as completed. This change removes this by getting rid of the `futureResponseMessage` capture, instead capturing the message id. This change also makes it so that we don't need to copy it into the lambda.


Differential Revision: [D22127398](https://our.internmc.facebook.com/intern/diff/D22127398/)

rohan-varma committed:
Remove variable shadowing in TensorPipe lambda

`futureResponseMessage` is shadowed in the `pipeWrite` lambda which
creates some confusion, also, this makes it so that we don't need to copy it
into the lambda

Differential Revision: [D21758911](https://our.internmc.facebook.com/intern/diff/D21758911/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Remove variable shadowing in TensorPipe lambda"


`futureResponseMessage` is shadowed in the `pipeWrite` lambda which
creates some confusion, since it is used in the initial error handling but then a future of the same name is created when marking the future as completed. This change removes this by getting rid of the `futureResponseMessage` capture, instead capturing the message id. This change also makes it so that we don't need to copy it into the lambda.


Differential Revision: [D21758911](https://our.internmc.facebook.com/intern/diff/D21758911/)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -337,8 +337,10 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
       std::make_shared<FutureMessage>();
+
   requestMessage.setId(nextMessageID_++);
-  pendingResponseMessage[requestMessage.id()] = futureResponseMessage;
+  auto requestMessageId = requestMessage.id();
+  pendingResponseMessage[requestMessageId] = futureResponseMessage;
```
Nit:
```
auto requestMessageId = nextMessageID_++;
requestMessage.setId(requestMessageId);
```
This may be more readable than first setting the id on the message then getting it back.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -373,11 +375,16 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
       std::move(requestMessage),
-      [this, &clientPipe, futureResponseMessage](
-          const tensorpipe::Error& error) {
+      [this, &clientPipe, requestMessageId](const tensorpipe::Error& error) {
```
Is there no other way to avoid the shadowing? By storing just the ID we must do a lookup which, although fast, wouldn't be needed if we already had the pointer. (As this is only relevant in the error codepath it's not really important... mostly a matter of style)

rohan-varma replied:
We can simply change the variable name, although I decided to take this approach to also get rid of the additional copy into the lambda.

lw reviewed (commented):
Could you rebase this? The stack I just landed changed this code quite a bit...

osalpekar reviewed (commented):
Given @lw's suggestion about the pointer lookup, it might be better to just change the name of the other `futureResponseMessage` in the `pipeRead` callback. Maybe to something like `responseFuture`?

rohan-varma committed:
Update on "Remove variable shadowing in TensorPipe lambda"


`futureResponseMessage` is shadowed in the `pipeWrite` lambda which
creates some confusion, since it is used in the initial error handling but then a future of the same name is created when marking the future as completed. This change removes this by getting rid of the `futureResponseMessage` capture, instead capturing the message id. This change also makes it so that we don't need to copy it into the lambda.


Differential Revision: [D21758911](https://our.internmc.facebook.com/intern/diff/D21758911/)

[ghstack-poisoned]

rohan-varma commented:
On 2nd thought, I think it may be worth trying to avoid the future copy here (since it happens always - regardless of error case or not), as it's only used in the error case, and I think in the error case paying the price of a map lookup should be ok. Any more thoughts on this change?

rohan-varma commented:
@lw Any thoughts on this change? Trying to close out some older stuff from my PR queue.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -426,7 +427,13 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
           }
-          markFutureWithError(std::move(futureResponseMessage), error.what());
+          auto it = clientPipe.pendingResponseMessage_.find(requestMessageId);
+          TORCH_INTERNAL_ASSERT(
+              it != clientPipe.pendingResponseMessage_.end(),
```
I don't know whether we should assert this, as it could in fact reasonably happen: if the pipeRead callback for the response to a previous request receives an error then it's going to mark all pending futures with that error and clear the map, which means that this code will not be able to find the future. But, in that case, it should just keep going rather than raising an exception.

rohan-varma replied:
I see, yeah I just saw the error handling code where we flush all messages belonging to a particular pipe in case of an error in the pipeRead callback.

rohan-varma replied:
Updated to remove this.

lw reviewed (commented):
Sorry, this had slipped off my radar. You're right that as this just occurs in the error case it's not a problem to do an extra lookup. Also, by not holding a shared_ptr to the response, we will not prevent it from being destroyed and its memory from being freed.

rohan-varma committed:
Update on "Remove variable shadowing in TensorPipe lambda"


`futureResponseMessage` is shadowed in the `pipeWrite` lambda which
creates some confusion, since it is used in the initial error handling but then a future of the same name is created when marking the future as completed. This change removes this by getting rid of the `futureResponseMessage` capture, instead capturing the message id. This change also makes it so that we don't need to copy it into the lambda.


Differential Revision: [D21758911](https://our.internmc.facebook.com/intern/diff/D21758911/)

[ghstack-poisoned]

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -616,7 +615,8 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
           }
-          markFutureWithError(std::move(futureResponseMessage), error.what());
+          auto it = clientPipe.pendingResponseMessage_.find(messageId);
+          markFutureWithError(it->second, error.what());
```
I think you'll still need to check whether the future is in the map before marking with an error (doing `it->second` if the iterator has reached the end is undefined behavior I believe):

```
if (it != clientPipe.pendingResponseMessage_.end()) {
    markFutureWithError(it->second, error.what());
}
```

rohan-varma replied:
I see, yeah this could have caused some unfortunate errors at runtime.

osalpekar reviewed (approved):
I'll stamp to unblock, but let's wait for the above change + tests. Thanks!

rohan-varma committed:
Update on "Remove variable shadowing in TensorPipe lambda"


`futureResponseMessage` is shadowed in the `pipeWrite` lambda which
creates some confusion, since it is used in the initial error handling but then a future of the same name is created when marking the future as completed. This change removes this by getting rid of the `futureResponseMessage` capture, instead capturing the message id. This change also makes it so that we don't need to copy it into the lambda.


Differential Revision: [D22127398](https://our.internmc.facebook.com/intern/diff/D22127398/)

[ghstack-poisoned]

lw reviewed (approved):
Looks good, thanks!
